src/Maestro/Maestro.Common/AzureDevOpsTokens/AzureDevOpsTokenProvider.cs

```diff
@@ -26,42 +28,123 @@ namespace Maestro.Common.AzureDevOpsTokens;
 
-        foreach (var credential in options.CurrentValue.ManagedIdentities)
+    public string GetTokenForRepository(string repositoryUrl)
+    {
+        Match m = AccountNameRegex.Match(repositoryUrl);
+        if (!m.Success)
         {
-            _tokenCredentials[credential.Key] = credential.Value == "system"
-                ? new ManagedIdentityCredential()
-                : new ManagedIdentityCredential(credential.Value);
+            throw new ArgumentException($"{repositoryUrl} is not a valid Azure DevOps repository URL");
         }
+
+        var account = m.Groups["account"].Value;
+        return GetTokenForAccount(account);
     }
 
-    public async Task<string> GetTokenForAccount(string account)
+    public async Task<string?> GetTokenForRepositoryAsync(string repositoryUrl)
     {
-        if (_options.CurrentValue.Tokens.TryGetValue(account, out var pat) && !string.IsNullOrEmpty(pat))
+        Match m = AccountNameRegex.Match(repositoryUrl);
+        if (!m.Success)
         {
-            return pat;
+            throw new ArgumentException($"{repositoryUrl} is not a valid Azure DevOps repository URL");
         }
 
-        if (_tokenCredentials.TryGetValue(account, out var credential))
+        var account = m.Groups["account"].Value;
+        return await GetTokenForAccountAsync(account);
+    }
+
+    private TokenCredential GetCredential(string account)
+    {
+        if (_accountCredentials.TryGetValue(account, out var credential))
         {
-            return (await credential.GetTokenAsync(new TokenRequestContext([AzureDevOpsScope]))).Token;
+            return credential;
         }
 
-        // We can also define just one MI for all accounts
-        if (_tokenCredentials.TryGetValue("default", out var defaultCredential))
+        if (_accountCredentials.TryGetValue("default", out var defaultCredential))
         {
-            return (await defaultCredential.GetTokenAsync(new TokenRequestContext([AzureDevOpsScope]))).Token;
+            return defaultCredential;
         }
 
         throw new ArgumentOutOfRangeException(
             $"Azure DevOps account {account} does not have a configured PAT or credential. " +
             $"Please add the account to the 'AzureDevOps.Tokens' or 'AzureDevOps.ManagedIdentities' configuration section");
     }
+
+    private static Dictionary<string, TokenCredential> GetCredentials(
+        AzureDevOpsTokenProviderOptions options,
+        Func<string, TokenRequestContext, CancellationToken, string> patResolver)
+    {
+        Dictionary<string, TokenCredential> credentials = [];
+
+        foreach (var pair in options)
+        {
+            var account = pair.Key;
+            var option = pair.Value;
+
+            // 1. AzDO PAT from configuration
+            if (!string.IsNullOrEmpty(option.Token))
+            {
+                credentials[account] = new ResolvingCredential((context, cancellationToken) => patResolver(account, context, cancellationToken));
```

Comment:
Why are we resolving the credential each time even though it does not change throughout runtime? Wouldn't it be better to just use `ResolvedCredential`?

Reply:
But what if the service configuration changes? We need to resolve it again from the `IOptionsMonitor` to get the new value